metrics_sdk/lib/opentelemetry/sdk/metrics/state/metric_stream.rb

```diff
@@ -43,21 +44,15 @@ def collect(start_time, end_time)
                 @instrument_kind,
                 @meter_provider.resource,
                 @instrumentation_library,
-                @data_points.dup,
+                @aggregation.collect(start_time, end_time),
                 start_time,
                 end_time
               )
             end
           end
 
-          def update(measurement, aggregation)
-            @mutex.synchronize do
-              @data_points[measurement.attributes] = if @data_points[measurement.attributes]
-                                                       aggregation.call(@data_points[measurement.attributes], measurement.value)
-                                                     else
-                                                       measurement.value
-                                                     end
-            end
+          def update(value, attributes)
+            @mutex.synchronize { @aggregation.update(value, attributes) }
```

Comment:
Ah - is this what protects multiple callers from colliding when we update during aggregation?

Reply:
👍 it protects against concurrent reads and writes during `update` and `collect`. It does not protect against writes to metric data points while (metric) readers, such as exporters, are reading from the data points returned by `collect`. This happens when the aggregation temporality is not `:delta`. I think in that case, we should return a clone of each data point. The depth of cloning required will be interesting (e.g. we need to clone the buckets of a histogram data point, but not the boundaries or attributes).

Reply:
Will be resolved with suggestions elsewhere.